Firestore/core/src/firebase/firestore/nanopb/byte_string.cc

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/nanopb/byte_string.h"
+
+#include <cstdlib>
+#include <cstring>
+#include <ostream>
+
+#include "Firestore/core/src/firebase/firestore/nanopb/nanopb_util.h"
+#include "absl/strings/escaping.h"
+
+namespace firebase {
+namespace firestore {
+namespace nanopb {
+
+namespace {
+
+/**
+ * Creates a new, null-terminated byte array that's a copy of the given string
+ * value.
+ */
+pb_bytes_array_t* MakeBytesArray(const uint8_t* data, size_t size) {
+  pb_size_t pb_size = CheckedSize(size);
+
+  // Allocate one extra byte for the null terminator that's not necessarily
+  // there in a string_view. As long as we're making a copy, might as well
+  // make a copy that won't overrun when used as a regular C string. This is
+  // essentially just to make debugging easier--actual user data can have
+  // embedded nulls so we shouldn't be using this as a C string under normal
+  // circumstances.
+  auto result = static_cast<pb_bytes_array_t*>(
+      malloc(PB_BYTES_ARRAY_T_ALLOCSIZE(pb_size) + 1));
+  result->size = pb_size;
+  memcpy(result->bytes, data, pb_size);
+  result->bytes[pb_size] = '\0';
+
+  return result;
+}
+
+pb_bytes_array_t* MakeBytesArray(const char* data, size_t size) {
+  return MakeBytesArray(reinterpret_cast<const uint8_t*>(data), size);
+}
+
+}  // namespace
+
+ByteString::ByteString(const std::vector<uint8_t>& value)
+    : bytes_(MakeBytesArray(value.data(), value.size())) {
+}
+
+ByteString::ByteString(const pb_bytes_array_t* bytes)
```

Comment:
This ctor isn't declared in the header? I don't think you need it.

Reply:
Sorry, this bled through from a later commit state where it is needed. Removed.